File: demo_03_tworecon_dosecurve_LCD.m
% demo_04_running_comparisons.m
% authors: Brandon Nelson, Rongping Zeng
%
% This demo outputs AUC curves of two recon options to show how the LCD-CT tool can be used to compare to denoising devices or recon methods
%% add relevant source files and packages
addpath(genpath('src'))
% clear all;
% close all;
clc;

if is_octave
  pkg load image tablicious
end
%% User specified parameters
% specify the `base_directory` containing images to be evaluated
if ~exist('use_large_dataset', 'var')
    use_large_dataset = false;
end

observers = {LG_CHO_2D()};
% observers = {DOG_CHO_2D()}
% observers = {LG_CHO_2D(),...
%              DOG_CHO_2D(),...
%              GABOR_CHO_2D(),...
%              };

%% Select datasets
% download if necesarry
base_directory = 'data';
if use_large_dataset
    fullfile(base_directory, 'large_dataset');
    if ~exist(large_dataset_directory, 'dir')
        download_largedataset(base_directory)
    end
    base_directory = large_dataset_directory;
else
    base_directory = fullfile(base_directory, 'small_dataset');
end
recon_1_dir = fullfile(base_directory, 'fbp');
recon_2_dir  = fullfile(base_directory, 'DL_denoised');
%% Next specify a ground truth image
% This is used to determine the center of each lesion for Location Known Exactly (LKE) low contrast detection

ground_truth_fname = fullfile(base_directory,'fbp','ground_truth.mhd');
offset = 1000;
ground_truth = mhd_read_image(ground_truth_fname) - offset; %need to build in offset to dataset

%% run
nreader = 10;   
pct_split = 0.6
seed_split = randi(1000, nreader,1);
doses = dir(fullfile(recon_1_dir, 'dose_*'));
recon_1_res = [];
for i = 1:length(doses)
    dose_dir = fullfile(doses(i).folder, doses(i).name);
    res = measure_LCD(dose_dir, observers, ground_truth, offset, nreader, pct_split, seed_split);
    if is_octave
        res.recon = "fbp";
        if isempty(recon_1_res)
            recon_1_res = res;
        else
            recon_1_res = vertcat(recon_1_res, res);
        end
    else
        res.recon(:) = "fbp";
        recon_1_res = [recon_1_res; res];
    end
end


doses = dir(fullfile(recon_2_dir, 'dose_*'));
recon_2_res = [];
for i = 1:length(doses)
    dose_dir = fullfile(doses(i).folder, doses(i).name);
    res = measure_LCD(dose_dir, observers, ground_truth, offset, nreader, pct_split, seed_split);
    if is_octave
        res.recon = "DL denoised";
        if isempty(recon_2_res)
            recon_2_res = res;
        else
            recon_2_res = vertcat(recon_2_res, res);
        end
    else
        res.recon(:) = "DL denoised";
        recon_2_res = [recon_2_res; res];
    end
end
%% combine results
if is_octave
  res_table = vertcat(recon_1_res, recon_2_res);
else
  res_table = cat(1, recon_1_res, recon_2_res);
end
%% save results
fname = mfilename;
output_fname = ['results_', fname(1:7), '.csv'];
write_lcd_results(res_table, output_fname)
%% plot results
set_ylim = [];
if use_large_dataset
    set_ylim = [0.7 1.01];
end
if ~use_large_dataset
    warning("`use_large_dataset` (line 31) is set to false`. This script is using a small dataset (10 repeat scans) to demonstrate usage of the LCD tool. For more accurate results, set `use_large_dataset = true`")
end
figure('NumberTitle', 'off', 'Name', 'AUC vs. Dose Curves');
plot_results(res_table, set_ylim)

res_table
%% plot the difference
% let's just look at a subset
figure('NumberTitle', 'off', 'Name', 'AUC Difference');
diff_auc = recon_2_res.auc - recon_1_res.auc;
diff_res = recon_1_res;
diff_res.auc = diff_auc;
diff_res.recon(:) = "DL denoised - fbp";
plot_results(diff_res)

diff_res

if ~use_large_dataset
    warning("`use_large_dataset` (line 15) is set to false`. This script is using a small dataset (10 repeat scans) to demonstrate usage of the LCD tool. For more accurate results, set `use_large_dataset = true`")
end

File: demo_03_tworecon_dosecurve_LCD.py
"""
Demo 03: Two Recon Dose Curve LCD Analysis
"""
from pathlib import Path
import numpy as np
import pandas as pd
from lcdct.LCD import measure_LCD, plot_results
from lcdct.utils import load_dataset, read_mhd

def main():
    base_dir = Path('data')
    use_large_dataset = False
    
    if use_large_dataset:
        base_directory = base_dir / 'large_dataset'
    else:
        base_directory = base_dir / 'small_dataset'
        
    recon_names = ['fbp', 'DL_denoised']
    
    # Ground Truth
    ground_truth_fname = base_directory / 'fbp' / 'ground_truth.mhd'
    offset = 1000
    if not ground_truth_fname.exists():
        print(f"Ground truth not found at {ground_truth_fname}")
        return
        
    ground_truth = read_mhd(str(ground_truth_fname)).astype(np.float32) - offset
    
    results_list = []
    observers = ['LG_CHO_2D']

    # Find common dose levels
    # Scan fbp dir for dose_*
    fbp_dir = base_directory / 'fbp'
    dose_dirs = sorted(list(fbp_dir.glob('dose_*')))
    
    for d_path in dose_dirs:
        try:
            dose_val = int(d_path.name.split('_')[1])
        except Exception:
            continue
            
        print(f"Processing Dose {dose_val}...")
        
        for recon in recon_names:
            recon_dir = base_directory / recon / d_path.name
            
            try:
                sp, sa = load_dataset(recon_dir, offset=offset)
            except Exception as e:
                # print(f"Skipping {recon_dir}: {e}")
                continue
            
            res = measure_LCD(sp, sa, ground_truth, observers=observers)
            res['recon'] = recon
            res['dose_level'] = dose_val
            results_list.append(res)

    if results_list:
        final_df = pd.concat(results_list, ignore_index=True)
        print("Results Head:")
        print(final_df.head())
        
        final_df.to_csv('results_demo_03.csv', index=False)
        plot_results(final_df, ylim=[0.5, 1.0])
    
if __name__ == '__main__':
    main()
